tests/test_blake3.py

```diff
@@ -182,3 +182,45 @@ def test_key_context_incompatible():
         pass
     else:
         assert False, "expected a type error"
+
+
+def test_name():
+    b = blake3()
+    assert b.name == "blake3"
+
+
+def test_copy_basic():
+    b = make_input(10**6)
+    b2 = make_input(10**6)
+    h1 = blake3(b)
+    expected = h1.digest()
+    h2 = h1.copy()
+    assert expected == h2.digest()
+    h1.update(b2)
+    expected2 = h1.digest()
+    assert expected2 != h2.digest(), "Independence test failed"
+    h2.update(b2)
+    assert expected2 == h2.digest(), "Update state of copy diverged from expected state"
+
+
+def test_copy_multithreading():
```

Comment:
Ah seeing this test, I think I can clear something up. The `multithreading` flag is scoped to a single call to `update`. So if you say `hasher.update(my_buf, multithreading=True)`, then potentially many threads will start looking at `my_buf`. But at the point when `update` returns, all those threads are done. They don't keep working in the background, and we don't need to worry about them interfering with future calls in some way.

There are a couple of reasons it works this way. As you've noticed, the alternative would be a safety nightmare. In practice, there's no way safe Rust would let us write that code, and all of the multithreading details in `blake3` are done with safe Rust. (The nuts and bolts are handled by the esteemed `rayon` crate.) What might arguably be worse than races against the hasher, would be races against the input buffer. If background threads were still reading the buffer after `update` returned, there would be absolutely no way for a caller to know when it was safe to reuse that buffer. Practically speaking, the hasher would need to take ownership of the buffer, and in Rust terms that would probably mean making an allocation internally and copying the caller's bytes into there. Of course that's overhead that we wouldn't want to pay.

So yes, the `multithreading` flag isn't as scary as it might sound, and we probably don't need a separate test for it. Also in this particular case, it actually has no effect, because we're not passing in any input bytes during construction. Basically this:

```python
my_buf = b"foo"
h1 = blake3(my_buf, multithreading=True)
```

is equivalent to this

```python
my_buf = b"foo"
h1 = blake3()
h1.update(my_buf, multithreading=True)
```

So if the buffer argument is empty or omitted, the `multithreading` argument doens't do anything. In particular, it's not retained as internal state.

Reply:
Good point, it shouldn't behave differently in a single (python) threaded context. I might have been getting mixed up with the GIL release aspect. 

The thing is, it *shouldn't* behave differently in the multithreading context, but suppose it did? I know that's super low probability with rust and all, but weird things happen. But I guess `test_multithreading` ought to catch it. Unless it doesn't. (I've been doing a lot of chaos engineering lately and it makes one a bit...paranoid ;) )

Reply:
It certainly doesn't do any harm to test it :) But it might be worth a comment at the top of the test to the effect of "This is not expected to catch any simple mistakes. If this test fails, but the other one passes, something truly weird is going on."